### PR TITLE
feat:  web3 notifications

### PR DESCRIPTION
**Pull request by ChefBingbong**
### Summary
📲🔗🗄️

This pull request adds a notifications feature to the web app using WalletConnect v2 and a PostgreSQL database. It updates the dependencies, environment variables, and WalletConnect integration in `apps/web/package.json`, `apps/web/.env`, and `apps/web/src/utils/wagmi.ts`. It adds new files for the WalletConnect DID document, push notification configuration, chain data context, WalletConnect client and push client context, notifications page, notifications view, and default view components in `apps/web/public/.well-known`, `apps/web/src/contexts`, `apps/web/src/pages`, and `apps/web/src/views`. It also adds new API routes for subscribing, unsubscribing, and handling the webhook events in `apps/web/src/pages/api/notifications`. Finally, it adds a new database schema file using drizzle-orm in `apps/web/src/db/schema.ts`.

> _Oh, we're the wagmi crew and we're sailing on the web_
> _We're using WalletConnect v2 to notify our friends_
> _We've got a PostgreSQL database to store our subs_
> _And we're heaving on the ropes with drizzle-orm and lokijs_

### Walkthrough
*  Add environment variables, dependencies, and files for WalletConnect v2 and notifications feature ([link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-38ac21448bcc34d3cfc05cb4c44552b636d08f9008fe911181e363c4af0be4a7R5-R6), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8L54-R69), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8R78-R79), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8L78-R97), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8R151), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-8fb20c287d9f64663eea50168ecea60f96fd71c9b6890de5d27ec846f5aa6f8fL1-R20), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-50786300b234b9dd539dbcfeaa62e5eea2d31758abc91af52cdec4c8d0d6c6d7L1-R26), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-1cc97eed41c277697e8f4e6c25d99c08735b2995913b38eaf9d9d78229f69bf4R1-R9), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-1677e68485cb8637ddb15fee5d910543a96cd8b2c1b18da418a4caf84e5f9fd8R49-R72), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-1677e68485cb8637ddb15fee5d910543a96cd8b2c1b18da418a4caf84e5f9fd8L59-R83), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-b746ecf7aeb22ba05a0775d1f2dfe06c2a8060b11a267f6052829f5ff4f87185R21-R27), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-b746ecf7aeb22ba05a0775d1f2dfe06c2a8060b11a267f6052829f5ff4f87185L64-R86), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-b746ecf7aeb22ba05a0775d1f2dfe06c2a8060b11a267f6052829f5ff4f87185L120-R131), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-dee8e400585307d38fc00d235ed19fa72f81a39b6d338f97a9dc3f9cfd8d2872R1-R129), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-660f649ee76e0f1ff402ff7a8c20accca6d9398456d778a1e5d2bdb809c2accaR1-R68))
*  Remove unused dependencies and duplicates from package.json ([link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8L101-R127), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8L150-L151))
*  Wrap app component with WalletConnect client and push client context provider ([link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-f369940b97496422b1677fd27e878e921f282bfffbfdbd0b0607db7a9e6e2584R1-R262), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-a3569d521976a5ee9e97e2b2e005bf7423bc96bb49397b3adfb09cde5862bf93R26), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-a3569d521976a5ee9e97e2b2e005bf7423bc96bb49397b3adfb09cde5862bf93L83-R87), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-a3569d521976a5ee9e97e2b2e005bf7423bc96bb49397b3adfb09cde5862bf93L100-R105))
*  Create React context and provider for chain data of notifications feature ([link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-733bdb8e55dc57b7caaa4a12de4d17d4aa51aa140ac5e21b3a0ad50d86262a4bL1-R81))
*  Create Next.js API routes for subscribing, unsubscribing, and handling webhook events for notifications feature ([link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-472f10614ac32a50b86dc4749dbf842efd03aecb9749a2ef4dbb337d44fc9c47L1-R365), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-3d0bd5fbc7a92c0fab00904644d761411f4794de3a24505d99df51dd8b4745ffR1-R46), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-2e880d388718279eda01a4bd13dfcb78cc3e791b8924fc08848225562a2ac648R1-R43), [link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-566a113b20b0f35c571bd32a17d606b715fe96406aedbdfab6c99b2326c1b0a9R1-R85))
*  Create notifications page component and specify supported chains ([link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-6bcf1f1035f4dfb0fa17712edabd2c33ce23bddd0eb3e2965b69ddda427dcf1cR1-R11))
*  Create default view component and render sign in and sign buttons for WalletConnect ([link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-dee8e400585307d38fc00d235ed19fa72f81a39b6d338f97a9dc3f9cfd8d2872R1-R129))
*  Create chain data and metadata module for notifications feature ([link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-660f649ee76e0f1ff402ff7a8c20accca6d9398456d778a1e5d2bdb809c2accaR1-R68))
*  Add menu item for notifications page and disable it if user is on wrong network or not on Binance Smart Chain ([link](https://github.com/pancakeswap/pancake-frontend/pull/7335/files?diff=unified&w=0#diff-65aa38e2318ced313d1604e387ea9f7ff3017b805fe530a27bd99bbdb2a95c27R59-R61))

### CODE REVIEW

File: apps/web/src/pages/api/notifications/webhook.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+          .status(500)
+          .json({ success: false, message: error });
+      // }
+    }
+  } else if (event === "unsubscribed") {
+    try {
+      await db.delete(gmUsers).where(eq(gmUsers.account, account));
+      return res.status(200).json({ success: true });
+    } catch (error) {
+      console.error(`Unsubscribing account ${account} failed: `, error);
```
[the remark this answers is not in the document]
Comment (ChefBingbong) [reply apps/web/src/pages/api/notifications/webhook.ts:70]: intrntionallly pushed without clean. playing around with different methods of creating custom extened wc connector that supports push client. theres a few ways ive beeen trying which is why theres a lot of duplicated/commented code

1) extend wagmi create custom connector
2) write custom ethers provider (universal provider) using sign-client combined with push (requires implement own json rpc spec, so this would make wagmi redundant)
3) use current wagmi and then implement the pushclient in an extended context provider, ...ok but requires second sig

[the remark this answers is not in the document]
Comment (ChefBingbong) [reply apps/web/src/pages/api/notifications/webhook.ts:70]: also lot of new packages installed. tbd which we actually need

File: apps/web/src/utils/WalletConnectConnector2.ts
Comment (chef-jojo) [inline apps/web/src/utils/WalletConnectConnector2.ts:1]: Looks like we can simply extend the class with additional method, 

with listeners, we can extend the connect and disconnect method

```
class A extends WalletConnectConnector {
  async connect() {
    super.connect()
    this.#setupPushListeners()
  }
}
```

Comment (chef-jojo) [reply apps/web/src/utils/WalletConnectConnector2.ts:1]: wait, did you change anything in [EthereumProvider2.ts](https://github.com/pancakeswap/pancake-frontend/pull/7335/files#diff-1869746da48ee4f979ea1d7a2ad74e28dfa99952ac997ab867726337bfeded54)

Comment (chef-jojo) [reply apps/web/src/utils/WalletConnectConnector2.ts:1]: Ah I understand what you are doing.. I don't think fork it will last long as we have to keep catch up the upstream EthereumProvider..

File: apps/web/.env [hunk 2]
```diff
@@ -2,3 +2,5 @@ NEXT_PUBLIC_MM_API_URL=https://linked-pool.pancakeswap.com/quote-service
 NEXT_PUBLIC_QUOTING_API=https://swap-quoting.pancakeswap.com/quoting-service
 MOONPAY_API="https://api.moonpay.com"
 MOONPAY_PUBLISHABLE_KEY="pk_live_Ch5fat39X8NvMZwih2k7hK4sDrKanSPz"
+NEXT_PUBLIC_PROJECT_ID="ae5413feaf0cdaee02910dc807e03203"
+DB_POOLING_URL="postgres://postgres:Emcg45245!!@db.wvmovalpevndzwntfwcy.supabase.co:6543/postgres"
```
Comment (chef-jojo) [inline apps/web/.env:6]: remove?

File: apps/web/public/.well-known/did.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+      "controller": "did:web:web-git-feat-web3-notifications.pancake.run",
+      "publicKeyJwk": {
+        "kty": "OKP",
+        "crv": "X25519",
+        "x": "BRpWvDZPBr9ZAAHdqkYO5uhZATwfbVoJ6tgRwd1Znko"
+      }
+    }
+  ],
+  "keyAgreement": ["did:web:web-git-feat-web3-notifications.pancake.run#key-0"]
+}
```
Comment (chef-jojo) [inline apps/web/public/.well-known/did.json:17]: is this file necessary for wc?

**Comment by ChefBingbong** [comment]
yeah. this file was required by the wallet client to do verification and
and to establish connection between app and their BE. however they no
longer use the dapp push client. so maybe this isnt needed anymore. ill
will ping and get back to you. they also removed the page explaining the
reason for did.json from the docs too

File: apps/web/next.config.mjs [hunk 4]
```diff
@@ -63,6 +63,7 @@ const config = {
   reactStrictMode: true,
   swcMinify: true,
   images: {
+    domains: ['i.imgur.com', 'assets.pancakeswap.finance', 'pancakeswap.finance'],
```
Comment (chefjackson) [inline apps/web/next.config.mjs:66]: what's this new image domain? Can we move the asset we need to our codebase?

File: pnpm-lock.yaml
Comment (chefjackson) [inline pnpm-lock.yaml:1]: Pls downgrade the pnpm version to v8.6.0 and install again. New version of pnpm is causing module duplicate trouble

File: apps/web/.env [hunk 5]
```diff
@@ -1,3 +1,6 @@
 NEXT_PUBLIC_MM_API_URL=https://linked-pool.pancakeswap.com/quote-service
 NEXT_PUBLIC_QUOTING_API=https://swap-quoting.pancakeswap.com/quoting-service
-MOONPAY_API="https://api.moonpay.com"
\ No newline at end of file
+MOONPAY_API="https://api.moonpay.com"
+MOONPAY_PUBLISHABLE_KEY="pk_live_Ch5fat39X8NvMZwih2k7hK4sDrKanSPz"
+NEXT_PUBLIC_PROJECT_ID="ae5413feaf0cdaee02910dc807e03203"
+CAST_SERVER_SIGN_KEY='85a7792c-c5d7-486b-b37e-e752918e4866'
```
Comment (chefjackson) [inline apps/web/.env:6]: seems DEFAULT_CAST_SIGN_KEY isn't being used? Free to remove this env var?

File: apps/web/package.json [hunk 6]
```diff
@@ -51,10 +51,16 @@
     "@pancakeswap/v3-sdk": "workspace:*",
     "@pancakeswap/wagmi": "workspace:*",
     "@reduxjs/toolkit": "^1.9.1",
-    "@tanstack/react-query": "^4.28.0",
     "@sentry/nextjs": "^7.52.1",
     "@snapshot-labs/snapshot.js": "^0.4.40",
+    "@tanstack/react-query": "^4.28.0",
+    "@types/pg": "^8.10.2",
```
Comment (chefjackson) [inline apps/web/package.json:57]: pls move types to devDependencies

File: apps/web/src/components/FiatOnRampModal/FiatOnRampModal.tsx [hunk 7]
```diff
@@ -333,7 +334,7 @@ export const FiatOnRampModal = memo<InjectedModalProps & FiatOnRampProps>(functi
         )}
       </ModalWrapper>
       <Script
-        src="https://widget.mercuryo.io/embed.2.0.js"
+        src="https://sandbox-widget.mrcr.io/embed.2.0.js"
```
Comment (chefjackson) [inline apps/web/src/components/FiatOnRampModal/FiatOnRampModal.tsx:337]: pls use env variable if we need different widget js for different environments

**Comment by chefjackson** [comment]
Close as we have new pr
